### ISSUE TITLE
Feature-request: Output delay. Only the final output on the screen

### ISSUE THREAD

**Issue by ossossosso** [openstenoproject/plover#888]
This is an example of entries into a dictionary (Italian Language):

"Ch*a": "comunità",
"RO/PEa": "europea",
"Ch*a/RO/PEa": "Comunità Europea",

If I type separately strokes "Ch*a" and "RO/PEa" into a sentence, the output is lower case.

But if I type one next the other, the meaning is different, so the output should be Upper Case. And it works correctly, because longer strokes have priority.

This is what happens if I type "Ch*a/RO/PEa":
1) once pressed "Ch*a" the output is "comunità".
2) once pressed "RO" the output is "comunità ro".
3) once pressed "PEa", the flashing cursor comes back and delete "comunità ro", and replaces with "Comunità Europea".

This is one example, but there are other situations similar. 
The concept is always the same: while typing, the "event" of delete/replace happens often.

Perhaps it is an event limited to the only Italian language. I don't know.

It causes these issues:

1) For those who use a screen reader to listen the output, is a bit confusing. Audio Speech is not fluent and smooth. So you have the impression something wrong happens, and to be sure the output is correct, you have to stop and check manually.

2) Similar to the previous one. Word Processor application, in case of a mistake, underlines the wrong word. In this case, at the same time of underscore, the screen reader produces a sound, a sort of "bip", to inform the user of the mistake.
The event of "delete/replace" can cause some temporary mistakes before of the final output. This cause the word processor to underlines for a while and the screen reader to produces the sound. This adds more confusion, because you have to stop and check, even if finally there is no mistakes.

Do you think this can be managed in some way?

I've used "Output Delay" like title, because I've thought about an option to set an interval of time before to have final output on the screen, during this time Plover can "hide" the even of "delete/replace" and give directly the real output. Of course faster I type, lower I can set the delay.

Thank you so much for reading this post.

**Comment by benoit-pierre** [comment]
OK, so let's take one example, `SKEL/TOPB` will results in the following output:
```
SKEL -> "cancel"
TOPB -> "" -> "skeleton"
```

I assume the goal is for only one string being sent in this case, "skeleton".

Here's a [POC for that](https://github.com/openstenoproject/plover/compare/master...benoit-pierre:new_melani+delayed_output) (Windows build [here](https://ci.appveyor.com/project/benoit-pierre/plover/build/1733)).

Expected behavior with a delay of 1 second and the following sequences:
- at 0 second, `SKEL` is stroked, half a second later `TOPB` is stroked: at 1.5s "skeleton" is sent
- at 0 second, `SKEL` is stroked, 1.1 second later `TOPB`: at 1s "cancel" is sent, at 2.1s it's replaced by "skeleton"

Output is also flushed on key combos (and eventually the code will flush on commands too).

The output mode can be changed in the configuration output tab. Default delay is 0.5 second. No way to change it with the GUI yet, but it can be changed throught the configuration file:
```config
[Delayed Keyboard Emulation]
delay = 0.25
```

**Comment by morinted** [comment]
Would it be possible to send immediately if the output is not part of a
longer stroke?

**Comment by benoit-pierre** [comment]
No, not with the way the dictionaries work.

**Comment by benoit-pierre** [comment]
Or rather, we can't know if a series of stroke is a prefix of a longer series or not, but we can detect the start of a new series. Except we can do that in the translator, and there's no connection between the translator and the output stage itself...

**Comment by tkpwaeub** [comment]
How about a stroke that suspends translation in such a way that when you "unsuspend" you get the finished product of everything you did after you suspended translation? Would that be doable?

**Comment by morinted** [comment]
With retrospective actions, this would also be impossible, so I suppose a
delay is the more foolproof solution.

**Comment by morinted** [comment]
W.r.t. the auto send, not Gabe's suggestion.

**Comment by tkpwaeub** [comment]
Maybe it could just be a preference for what you mean by "toggle".

**Comment by tkpwaeub** [comment]
Or maybe a pop up "scratch pad" that you could focus on

**Comment by tkpwaeub** [comment]
I think some combination of the "repeat last word" plugin and something similar to the quarantine concept that Mirabai uses for vim captioning should do the trick??

**Comment by ossossosso** [comment]
I thank you very much for your fast reply to this request.

There are two links on the Benoit Poerre's post:

"Here's a POC for that (Windows build here)."

They contain the plugin for this kind of features? In case I don't know how to use.

Thank you so much.

**Comment by benoit-pierre** [comment]
Here's a newer build: https://ci.appveyor.com/project/benoit-pierre/plover/build/1736/artifacts

@ossossosso: it's a build with Melani and the delayed output integrated, just change the output type to `Delayed Keyboard Emulation` in the configuration's output tab. As explained above the default delay is 0.5 second and you'll have to manually edit Plover's configuration file if you want to change the value.

**Comment by ossossosso** [comment]
Thank you very much Benoit for this build.

It seems the delay can be set correctly. But basically the final effect is the same as before:

This is what happens if I type "Ch*a/RO/PEa":

1. once pressed "Ch*a" the output is "comunità".
2. once pressed "RO" the output is "comunità ro".
3. once pressed "PEa", the flashing cursor comes back and delete "comunità ro", and replaces with "Comunità Europea".

So the event "delete/replace" is still there. The only difference is that it happens with a delay of the seconds I've set.

**Comment by benoit-pierre** [comment]
Of course, in this case,  "comunità" is sent, "ro" is not (delayed because of the "Comunità Europea" replacement), but this still result in the replacement of the initial "comunità". I guess a more complex algorithm is needed.

**Comment by benoit-pierre** [comment]
Try this version instead: https://ci.appveyor.com/project/benoit-pierre/plover/build/1737/artifacts

**Comment by benoit-pierre** [comment]
The algorithm is more complex, basically output should end up split into words and delayed, so using a prefix stroke will further delay the output of the current word. Similarly, erasing/replacing text will also further delay output.

**Comment by ossossosso** [comment]
Thank you again for this last version.

I perfectly understand the complexity of it. A standard user like me is only able to see the outer part of the body, but I can only imagine the amount of gears inside, that have to work all together at the same time. So I thank you to consider my request.

The last version seems to have the same effect of the previous one: the event of "delete/replace" is still there, with a bit of delay, I don't notice any difference.

I don't know if it can help, but I share with you the way I see the issue, on the paper.

Standard Plover (keyboard emulation) works in two moments:
1. Press keys on the keyboard.
2. Output on the screen.

Your feature adds a third step (Delayed keyboard emulation), and you inserted in between:
1. Press keys on the keyboard.
2. Delay.
3. Output on the screen.

I think to get the result, the third step should be splitted into two differents:
3. Output not on the screen.
4. Output on the screen.

And the Delay feature can be positioned in between.
1. Press keys on the keyboard.
2. Output not on the screen.
3. Delay.
4. Output on the screen.

The input for the second step, "Output not on the screen", is the stenomachine. The input for the last step, "Output on the screen", is not the stenomachine, but the second step. "Output not on the screen".

**Comment by benoit-pierre** [comment]
Just to make sure: do you use the default value for the delay?

**Comment by ossossosso** [comment]
I didn't touch the plover.cfg file. I didn't add 
[Delayed Keyboard Emulation]
delay = 0.25

So I suppose the delay I see is the default value.
To be sure, before to unzip and run your last version, I've also deleted everything under the folder:
C:\Users\Username\AppData
regarding plover.

So I believe once run your version, it's like to run a new fresh installation of it.

**Comment by benoit-pierre** [comment]
And you set the output type to "Delayed Keyboard Emulation" in the configuration? How fast do you stroke everything?

**Comment by ossossosso** [comment]
Exactly, I've set the output type to "Delayed Keyboard Emulation" in the configuration. 
I've tried to type with a slow, medium and fast speed. The result is the same.

I've also tried to modify the plover.cfg file:
[Delayed Keyboard Emulation]
delay = 2.00

And type again with a slow, medium, fast speed. The result is again the same.

**Comment by benoit-pierre** [comment]
I run the following simulation:
* delay: 2 seconds
* minimal dictionary:
```json
    "Ch*a": "comunità",
    "RO": "ro",
    "Ch*a/RO/PEa": "Comunità Europea",
```
* strokes:
  * `Ch*a`
  * wait half a second, `RO`
  * wait half a second, `PEa`

and the output is correct: only "  Comunità Europea" get sent.

**Comment by benoit-pierre** [comment]
@ossossosso: do you have space placement set to "after"? That would explain the different behavior were seeing.

**Comment by benoit-pierre** [comment]
Third time the charm: https://ci.appveyor.com/project/benoit-pierre/plover/build/1741/artifacts

This version should work with both space placements.

**Comment by ossossosso** [comment]
Wow... it seems to work perfectly.
I should test for a working day before to confirm.

But... ehm... You know, the blanket is always short, now the feet are out :)

One really useful feature of Plover is the ability to simulate QWERTY keyboard keys, for example I use a lot "CTRL+V" to paste clipboard 
"#C": "{#Control_L(v)}{#Space}",
and other QWERTY keyboard shortcut during the workflow.

With "Delay" feature ON, these kind of entries don't work.

**Comment by benoit-pierre** [comment]
Damn... I really need more tests, but all that threading/timing stuff makes it harder (I know, I know, excuses). New build: https://ci.appveyor.com/project/benoit-pierre/plover/build/artifacts

Another plugin idea: paste/cut/copy from/to clipboard metas.

**Comment by ossossosso** [comment]
My compliments, clap clap clap :)

It works.
Can be added to the plugin too if other want to try.

Thank you very much Benoiut.

**Comment by mkrnr** [comment]
Closing this ticket due to inactivity. Please comment if this is still an issue.
